File: haskell/src/Day5.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Avoid lambda" #-}
module Day5 (day5) where

import Data.List
import Data.List.Split
import Data.Maybe
import Data.Ord
import Debug.Trace
import System.IO
import Control.Monad

day5 = do
    contents <- readFile "input/day5.txt"
    let line_arr = lines contents
    let orderings = orders line_arr
    let updates = updatesList line_arr
    print (part1 orderings updates)
    print (part2 orderings updates)

part1 :: [(Int, Int)] -> [[Int]] -> Int
part1 orderings updates = sum (map (\l -> l !! div (length l) 2) (filter (isCorrect orderings) updates))

part2 :: [(Int, Int)] -> [[Int]] -> Int
part2 orderings updates = sum (map (\l -> fixUpdate orderings l !! div (length l) 2) (filter (not . isCorrect orderings) updates))

isCorrect :: [(Int, Int)] -> [Int] -> Bool
isCorrect orderings [] = True
isCorrect orderings update = all (\it -> (it, head update) `notElem` orderings ) (tail update) && isCorrect orderings (tail update)

-- credit to olive for this one
fixUpdate :: [(Int, Int)] -> [Int] -> [Int]
fixUpdate orderings = sortBy (\num1 num2 ->
    if (num2, num1) `elem` orderings then GT
    else if (num1, num2) `elem` orderings then LT
    else EQ)

orders :: [String] -> [(Int, Int)]
orders lines = map ((\l -> (read (head l), read (l !! 1))) . splitOn "|") (filter  ('|' `elem`) lines)

updatesList :: [String] -> [[Int]]
updatesList lines =  map (map read . splitOn ",") (filter  (',' `elem`) lines)

File: kotlin/app/src/main/kotlin/days/Day5.kt
package days.day5

import java.io.File

fun day5() {
    val chars = File("../../input/day5.txt").readText().lines()
    var orderings: MutableList<Pair<Int, Int>> = arrayListOf()
    var updates: MutableList<List<Int>> = arrayListOf()
    for (line in chars ) {
        if (line.contains('|')) {
            val split = line.split('|')
            val pair = Pair(split[0].toInt(), split[1].toInt())
            orderings.add(pair)
        }
        else if (line.contains(',')) {
            val split = line.split(',')
            updates.add(split.map { s -> s.toInt() })
        }
    }

    part1(orderings, updates)
    part2(orderings, updates)
}

fun part1(orderings: MutableList<Pair<Int, Int>>, updates: MutableList<List<Int>>) {
    val correct = updates.filter { update -> isCorrect(orderings, update) }
    val sum = correct.map { list -> list[list.size / 2] }.sum()
    println(sum)
}

fun part2(orderings: MutableList<Pair<Int, Int>>, updates: MutableList<List<Int>>) {
    val incorrect = updates.filter { update -> !isCorrect(orderings, update) }
    val sum = incorrect.map { list -> fix(orderings, list)[list.size / 2] }.sum()
    println(sum)
}

fun isCorrect(orderings: MutableList<Pair<Int, Int>>, update: List<Int>): Boolean {
    for ((i, item) in update.withIndex()) {
        val pairs = update.subList(i + 1, update.size).map { it -> orderings.contains(Pair(it, item)) }
        if (pairs.any { it }) { return false }
    }

    return true
}

fun fix(orderings: MutableList<Pair<Int, Int>>, update: List<Int>): List<Int> {
    return update.sortedBy { 
        elem -> update.minus(elem).filter {
            num -> orderings.contains(Pair(num, elem))
        }.size
    }
}